src/rpc/client.rs

```diff
@@ -258,193 +118,259 @@ impl RpcClient {
-                    })
-                    .collect();
-
-                future::try_join_all(tasks)
+        self.clone()
+            .group_tasks_by_region(keys)
+            .try_for_each(move |(region_id, keys)| {
+                let cf = cf.clone();
+                self.clone()
+                    .raw_from_id(region_id)
+                    .and_then(|context| context.client.raw_batch_delete(context.region, cf, keys))
             })
-            .map_ok(|_| ())
     }
 
     pub fn raw_scan(
-        &self,
+        self: Arc<Self>,
         range: BoundRange,
         limit: u32,
         key_only: bool,
         cf: Option<ColumnFamily>,
     ) -> impl Future<Output = Result<Vec<KvPair>>> {
-        struct State {
-            limit: u32,
-            key_only: bool,
-            cf: Option<ColumnFamily>,
-        }
-
-        let scan: ScanRegionsContext<Vec<KvPair>, State> = ScanRegionsContext::new(
-            range,
-            State {
-                limit,
-                key_only,
-                cf,
-            },
-        );
-        let inner = Arc::clone(&self.inner);
-        loop_fn((inner, scan), |(inner, scan)| {
-            inner.locate_key(scan.start_key()).and_then(|location| {
-                let region = location;
-                let cf = scan.state.cf.clone();
-                inner
-                    .clone()
-                    .region_context_by_id(region.id())
-                    .map_ok(|(region, client)| {
-                        (scan, region.range(), RawContext::new(region, client, cf))
-                    })
-                    .and_then(|(scan, region_range, context)| {
-                        let (ref start_key, ref end_key) = scan.range;
-                        context
-                            .client()
-                            .raw_scan(
-                                context,
-                                Some(start_key.clone()),
-                                end_key.clone(),
-                                scan.state.limit,
-                                scan.state.key_only,
-                            )
-                            .map_ok(|pairs| (scan, region_range, pairs))
-                    })
-                    .map_ok(|(mut scan, region_range, mut pairs)| {
-                        let limit = scan.state.limit;
-                        scan.result.append(&mut pairs);
-                        if scan.result.len() as u32 >= limit {
-                            Loop::Break(scan.result)
-                        } else {
-                            match scan.next(region_range) {
-                                ScanRegionsStatus::Continue => Loop::Continue((inner, scan)),
-                                ScanRegionsStatus::Break => Loop::Break(scan.result),
-                            }
-                        }
-                    })
+        self.regions_for_range(range)
+            .try_fold(Vec::new(), move |mut result, context| {
```

Comment:
I wonder if we can guessimate the required capacity for the vector to save reallocations?

Reply:
Given my experience with Prost, I'd be hesitant to do this before we can prove that it improves performance. I'm also not sure what that would look like, but perhaps there is a rough heuristic for how many regions we might expect for a given number of tasks?